docs/RFCS/replica_tombstone.md

```diff
@@ -0,0 +1,101 @@
+   range R, it starts an election.
+5. Nodes A and B see that node C has a more advanced log position for
+   range R than they do (since they have nothing), so they vote for it.
+   C becomes leader and sends snapshots to A and B.
+6. There are now two "live" versions of the range. Clients
+   (`DistSenders`) whose range descriptor cache is out of date may
+   talk to the ABC group instead of the correct DEF group.
+
+The problems caused by removing replicas are also discussed in   [#768](https://github.com/cockroachdb/cockroach/issues/768).
+
+# Detailed design
+
+A **Replica ID** is generated with every `ChangeReplicas` call and
+stored in the `Replica` message of the range descriptor. Replica IDs
+are monotonically increasing within a range and never reused. They are
+generated using a `next_replica_id` field which will be added to the
+`RangeDescriptor` (alternative generation strategies include using the
+raft log position or database timestamp of the `EndTransaction` call
+that commits the membership change, but this information is less
+accessible at the point where it would be needed).
+
+The raft transport will send the replica ID with every
+`RaftMessageRequest`. This is how the `Store` will learn of the replica
+ID for new ranges. The `Store` will drop incoming messages with a
+replica ID that is less than the last known one for the range (in
+order to minimize disruption from out-of-date servers).
+
+The `DistSender` will also send the replica ID in the header of every
+request, and the receiver will reject requests with incorrect replica
+IDs (This will be rare since a range will normally be absent from a
+node for a time before being re-replicated).
+
+When a replica is garbage-collected, we write a **tombstone**
+containing the replica ID, so that the store can continue to drop
+out-of-date raft messages even after the GC.
+
+In the scenario above, with this change node C would send node A and
+B's replica IDs in its vote requests. They would see that they have a
+tombstone for that replica and not recreate it.
+
+# Drawbacks
+
+Tombstones must be long-lived, since a node may come back online after
+a lengthy delay. We cannot garbage-collect tombstones unless we also
+guarantee that no node will come back after a period of time longer
+than the tombstone GC time.
+
+# Alternatives
+
+If we had an explicit replica-creation RPC instead of creating them
+automatically on the first sighting of a raft message, this problem
+may go away. However, doing so would be tricky: this RPC would need to
+be retried in certain circumstances, and it is difficult to
+distinguish cases where a retry is necessary from cases that lead to
+the problem discussed here.
+
+# Unresolved questions
+
+* How long must we wait before garbage-collecting tombstones? Can we
+  do it at all?
+* How will the raft transport get the correct replica ID? Will
+  `MultiRaft` itself have to be aware of replica IDs for proper
+  synchronization?
```

Comment:
Hmm, that's a possibility. Raft only really needs to know about the ReplicaID, and in fact if we give it replica IDs instead of store IDs we wouldn't even need tombstones. We'd just need `MultiRaft` to be able to map group + replica IDs to node + store IDs to route messages and coalesce heartbeats. I think a leader would always be able to query the `Store` to get the current members of the range and perform this mapping. The tricky thing would be routing responses for followers. We'd need to remember the source address of any request with a previously-unseen `From` (or be able to use our RPC connections bidirectionally). This at least deserves to be written up as an alternative.